File: 1-2-procedures-and-processes-they-generate/1-2-1-linear-recursion-and-iteration/factorial.rb
#!/usr/bin/env ruby

# SICP 1.2.1
# Linear Recursion and Interation# Iterative

def fact2(n)
  return fact2iter(1, 1, n)
end

def fact2iter(product, counter, maxcount)
  return product if counter > maxcount
  return fact2iter(counter*product, counter+1, maxcount)
end

File: 1-2-procedures-and-processes-they-generate/1-2-1-linear-recursion-and-iteration/factorial.py
#!/usr/bin/env python3

# SICP 1.2.1
# Linear Recursion and Iteration

def fact1(n: int) -> int:
    if n == 1:
        return 1
    else:
        return n * fact1(n - 1)

# Iterative

def fact2(n: int) -> int:
    return fact2Iter(1, 1, n)

    def fact2Iter(product, counter, maxCount: int) -> int:
        if counter > maxCount:
            return product
        else:
            return fact2Iter(counter*product, counter+1, maxCount)


print(fact1(7))
print(fact2(7))
